src/dbnode/storage/index_test.go

```diff
@@ -65,6 +71,133 @@ func TestNamespaceIndexCleanupExpiredFilesets(t *testing.T) {
 	require.NoError(t, idx.CleanupExpiredFileSets(now))
 }
 
+func TestNamespaceIndexCleanupDuplicateFilesets(t *testing.T) {
+	md := testNamespaceMetadata(time.Hour, time.Hour*8)
+	nsIdx, err := newNamespaceIndex(md, testShardSet, DefaultTestOptions())
+	require.NoError(t, err)
+
+	idx := nsIdx.(*nsIndex)
+	now := time.Now().Truncate(time.Hour)
+	indexBlockSize := 2 * time.Hour
+	blockTime := now.Add(-2 * indexBlockSize)
+
+	fset1, err := ioutil.TempFile("", "fileset-9000-0-")
+	require.NoError(t, err)
+	fset2, err := ioutil.TempFile("", "fileset-9000-1-")
+	require.NoError(t, err)
+	fset3, err := ioutil.TempFile("", "fileset-9000-2-")
+	require.NoError(t, err)
+
+	volumeType := "extra"
+	infoFiles := []fs.ReadIndexInfoFileResult{
+		{
+			Info: indexpb.IndexVolumeInfo{
+				BlockStart: blockTime.UnixNano(),
+				BlockSize:  int64(indexBlockSize),
+				Shards:     []uint32{0, 1, 2},
+				IndexVolumeType: &protobuftypes.StringValue{
+					Value: volumeType,
+				},
+			},
+			AbsoluteFilepaths: []string{fset1.Name()},
+		},
+		{
+			Info: indexpb.IndexVolumeInfo{
+				BlockStart: blockTime.UnixNano(),
+				BlockSize:  int64(indexBlockSize),
+				Shards:     []uint32{0, 1, 2},
+				IndexVolumeType: &protobuftypes.StringValue{
+					Value: volumeType,
+				},
+			},
+			AbsoluteFilepaths: []string{fset2.Name()},
+		},
+		{
+			Info: indexpb.IndexVolumeInfo{
+				BlockStart: blockTime.UnixNano(),
+				BlockSize:  int64(indexBlockSize),
+				Shards:     []uint32{0, 1, 2, 3},
+				IndexVolumeType: &protobuftypes.StringValue{
+					Value: volumeType,
+				},
+			},
+			AbsoluteFilepaths: []string{fset3.Name()},
+		},
+	}
+
+	idx.readIndexInfoFilesFn = func(
+		filePathPrefix string,
+		namespace ident.ID,
+		readerBufferSize int,
+	) []fs.ReadIndexInfoFileResult {
+		return infoFiles
+	}
+	idx.deleteFilesFn = func(s []string) error {
+		require.Equal(t, []string{fset1.Name(), fset2.Name()}, s)
+		multiErr := xerrors.NewMultiError()
+		for _, file := range s {
+			multiErr = multiErr.Add(os.Remove(file))
+		}
+		return multiErr.FinalError()
+	}
+	require.NoError(t, idx.CleanupDuplicateFileSets())
+}
+
+func TestNamespaceIndexCleanupDuplicateFilesetsNoop(t *testing.T) {
+	md := testNamespaceMetadata(time.Hour, time.Hour*8)
+	nsIdx, err := newNamespaceIndex(md, testShardSet, DefaultTestOptions())
+	require.NoError(t, err)
+
+	idx := nsIdx.(*nsIndex)
+	now := time.Now().Truncate(time.Hour)
+	indexBlockSize := 2 * time.Hour
+	blockTime := now.Add(-2 * indexBlockSize)
+
+	fset1, err := ioutil.TempFile("", "fileset-9000-0-")
+	require.NoError(t, err)
+	fset2, err := ioutil.TempFile("", "fileset-9000-1-")
+	require.NoError(t, err)
```

Comment:
Want to use a tempdir as first arg here just like prev suggestion (and defer removeall the tempdir)?